src/referenced-fields.ts

```diff
@@ -294,3 +322,302 @@ export const getModifiedFields: EngineInstance['getModifiedFields'] = (
+
+			break;
+
+		case 'In':
+		case 'NotIn':
+			// TODO: `NotIn` has no node type defined
+			const inNode = constraint as InNode;
+			for (const arg of inNode.slice(2)) {
+				assertAbstractSqlIsNotLegacy(arg);
+				findBindCandidates(arg, candidates);
+			}
+
+			break;
+
+		case 'Exists':
+		case 'NotExists':
+			const existsNode = constraint as ExistsNode | NotExistsNode;
+			const existentialInner = existsNode[1];
+			assertAbstractSqlIsNotLegacy(existentialInner);
+			findBindCandidates(existentialInner, candidates);
+
+			break;
+
+		case 'Not':
+			const notNode = constraint as NotNode;
+			const negationInner = notNode[1];
+			assertAbstractSqlIsNotLegacy(negationInner);
+			findTablesInConstraint(negationInner, candidates);
+
+			break;
+
+		case 'And':
+		case 'Or':
+			const andNode = constraint as AndNode | OrNode;
+			for (const arg of andNode.slice(1)) {
+				assertAbstractSqlIsNotLegacy(arg);
+				findTablesInConstraint(arg, candidates);
+			}
+
+			break;
+
+		default:
+			throw new Error(`unknown constraint type: ${constraint[0]}`);
+	}
+};
+
+const findTableInTableDefinition = (
+	statement: AliasableFromTypeNodes,
+	candidates: BindCandidate[],
+): AliasedTable | null => {
+	switch (statement[0]) {
+		case 'SelectQuery':
+		case 'UnionQuery':
+			findBindCandidates(statement, candidates);
+
+			return null;
+
+		case 'Table':
+			const tableNode = statement as TableNode;
+
+			return { tableName: tableNode[1], alias: tableNode[1] };
+
+		case 'Alias':
+			const aliasNode = statement as AliasNode<FromTypeNodes>;
+			const table = findTableInTableDefinition(aliasNode[1], candidates);
+			if (table) {
+				return { tableName: table.tableName, alias: aliasNode[2] };
+			}
+
+			return null;
+
+		default:
+			throw new Error(`unknown table definition type: ${statement[0]}`);
+	}
+};
+
+const findTableInStatement = (
+	statement: SelectQueryStatementNode,
+	candidates: BindCandidate[],
+): AliasedTable | null => {
+	switch (statement[0]) {
+		case 'Select':
+			// Select lists may contain subqueries so introspect into them as well
+			const selectNode = statement as SelectNode;
+			for (const arg of selectNode[1]) {
+				assertAbstractSqlIsNotLegacy(arg);
+				findBindCandidates(arg, candidates);
+			}
+
+			return null;
+
+		case 'From':
+		case 'CrossJoin':
+			const fromNode = statement as FromNode | CrossJoinNode;
+
+			return findTableInTableDefinition(fromNode[1], candidates);
+
+		case 'Join':
+		case 'LeftJoin':
+		case 'RightJoin':
+		case 'FullJoin':
+			const joinNode = statement as
+				| InnerJoinNode
+				| LeftJoinNode
+				| RightJoinNode
+				| FullJoinNode;
+			if (joinNode[2]) {
+				findTablesInConstraint(joinNode[2][1], candidates);
+			}
+
+			return findTableInTableDefinition(joinNode[1], candidates);
+
+		case 'Where':
+			const whereNode = statement as WhereNode;
+			findTablesInConstraint(whereNode[1], candidates);
+
+			return null;
+
+		default:
+			return null;
+	}
+};
+
+const findBindCandidates = (
```

Comment:
Some jsdoc detailing the intention of this function would be useful, I initially thought it would be looking for bind vars but it looks like it's actually looking for queries into which to add bind vars. And the same thing for the other functions - I have a feeling a bunch of them can be named more clearly and possibly collapsed but it's hard to say for sure right now as I don't fully understand all of the interactions